bleach/html5lib_shim.py

```diff
@@ -51,9 +200,55 @@ def consumeEntity(self, allowedChar=None, fromAttribute=False):
         else:
             self.tokenQueue.append({"type": tokenTypes['Characters'], "data": '&'})
 
+    def emitCurrentToken(self):
+        token = self.currentToken
+
+        if ((self.parser.tags is not None and
+             token['type'] in (tokenTypes['StartTag'], tokenTypes['EndTag']) and
+             token['name'].lower() not in self.parser.tags)):
+            # If this is a start/end tag for a tag that's not in our allowed
+            # list, then it gets stripped or escaped. In both of these cases
+            # it gets converted to a Characters token.
+            if self.parser.strip:
+                # If we're stripping the token, we just throw in an empty
+                # string token.
+                new_data = ''
+
+            else:
+                # If we're escaping the token, we want to escape the exact
+                # original string. Since tokenizing also normalizes data
+                # and this is a tag-like thing, we've lost some information.
+                # So we go back through the stream to get the original
+                # string and use that.
+                new_data = get_recent_tag_string(self.stream.stream_history(), token)
+
+            new_token = {
+                'type': tokenTypes['Characters'],
+                'data': new_data
+            }
+
+            self.currentToken = new_token
+            self.tokenQueue.append(new_token)
+            self.state = self.dataState
+            return
+
+        super(BleachHTMLTokenizer, self).emitCurrentToken()
+
 
 class BleachHTMLParser(HTMLParser):
     """Parser that uses BleachHTMLTokenizer"""
+    def __init__(self, tags, strip, **kwargs):
+        """
+        :arg tags: list of allowed tages--everything else is either stripped or
+            escaped; if None, then this doesn't look at tags at all
+        :arg strip: whether to strip disallowed tags (True) or escape them (False);
+            if tags=None, then this doesn't have any effect
```

Comment:
Would it make sense to have separate parsers for strip and escape modes?

Reply:
That's really interesting. It would make sense if it reduced the likelihood of mistakes and/or was more performant. I'll write up an issue to look into that after I land this.